rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,584 @@
+        the version of an Artifact equals to the _&lt;major>.&lt;minor>_ version
+        of its producer component. This version will be further used as
+        consumption hint for these Artifacts, to correctly handle version
+        upgrades.
+*   **No Forward compatibility guarantee on payload**: while the system will not
+    reject passing Artifacts produced from newer versions of TFX to an old
+    version of a component, we do not guarantee this will work. A warning will
+    be generated in this case.
+*   Resolvers and importers would only resolve Artifacts produced with same or
+    older versions of the library, by default. A warning will be issued for
+    any artifacts skipped due to newer versions.
+
+#### SDK (aka DSL), Official Components {#sdk-aka-dsl-official-components}
+
+All parts of SDK (aka DSL), IR, interfaces of official components shall comply
+with SemVer. The functionalities form the following core CUJs, which include but
+are not limited to:
+
+*   Defining or declaring components;
+*   Composing them into a pipeline graph;
+*   Environment/Deployment specific decoration of nodes;
+*   Advanced control semantics (condition/loop).
+
+It’s worth noting again that APIs from underlying packages used in above core
+CUJs shall also comply.
+
+#### Custom Component {#custom-component}
+
+There are several ways to define a Custom Component in TFX::OSS:
+
+1.  Importing a user provided python function
+    *   this process and all APIs involved will be backward compatible;
+2.  Importing a user provided container image
+    *   this process and APIs will be backward compatible;
+3.  Defining ComponentSpec, ExecutorSpec and subclassing with component class
+    *   This is a very complex user interface, and arguably exposes too much
+        internal implementation details. Moreover, it’s proven that this is
+        quite difficult to do for our users. A possible improvement proposed is
+        to combine the component and executor classes, and have the convention
+        be that users subclass a base component's class to extend or override
+        functionality. Until this is settled, this API will not comply with
+        backward compatibility.
+
+#### Customization on ExampleGen {#customization-on-examplegen}
+
+There are several ways to customize ExampleGen:
+
+1.  Providing a PTransform for converting input source to examples by inheriting
+    from BaseExampleGenExecutor and overwriting the abstract method.
+    *   This is recommended and will be backward compatible.
+2.  Use the above custom Component or custom Executor for use cases like join.
+    *   This is a very complex user interface, and arguably exposes too much
+        internal implementation details (driver:external file detection,
+        span/version/split etc). Moreover, it’s proven that this is quite
+        difficult to do for our users. We plan to mark this process as
+        experimental in 1.0, and not comply with backward compatibility.
```

Comment:
A bit worried on this here for 2 reasons: 

1) **ExampleGen dependency on external systems**
ExampleGen is the main component that depends (implicitly or explicitly) on external systems from TFX. A lot of times you will need some custom code to check on external systems or do some custom metadata preparation before actually running the Executor. This change here, means that the only way to do that is via an experimental API. 

Alternatives:
- Pushing this logic to the scheduler/orchestrator: Feasible but reduces flexibility/portability of TFX pipeline
- Pushing this logic to Executor: Marked as experimental in this RFC as well.

While most users will not use this, advanced users will require this customization.

2) Custom output data formats: As discussed, tf.example may not be the best format for all use cases (ex: recommender systems). As such, you would need to generate custom ExampleGens that output custom formats as needed. This can be done today by customizing the Excecutor on ExampleGen to change output format and using Data views.

Alternatives:
- Push this outside of TFX to external system: Loss of compatibility/portability + Still requires a custom component to import the dataset inside the TFX pipeline.

CC @aoen

Reply:
@1025KB